toolchain/check/merge.cpp

```diff
@@ -233,7 +233,14 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
 
   auto prev_param_type_id = SemIR::GetTypeInSpecific(
       context.sem_ir(), prev_specific_id, prev_param_pattern.type_id());
-  if (!context.types().AreEqualAcrossDeclarations(new_param_pattern.type_id(),
+  if ((!allow_self_type_mismatch ||
```

Comment:
This is a pretty complex `if`, maybe add a brief implementation comment?

e.g.:

```suggestion
  // Make sure the param type matches, unless it's a `self` param and a different type is allowed.
  if ((!allow_self_type_mismatch ||
```

Reply:
Now that the name check is pulled out separately, maybe not worth the comment? (at a similar level of complexity as the other checks in this function - and easier to read when it's an affirmative check "allow self type mismatch && new name == self && prev name == self, then return true/skip checking the type")